File: src/Cell.ts
import { Coordinate } from "./Coordinate";

export type AliveCell = {
    kind: "Alive",
    Coordinate: Coordinate,
}

export type DeadCell = {
    kind: "Dead",
    Coordinate: Coordinate,
}

export type Cell = AliveCell | DeadCell;

function isDead(cell: Cell): cell is DeadCell {
    return cell.kind === "Dead";
}

export function isAlive(cell: Cell): cell is AliveCell {
    return cell.kind === "Alive";
}

export function checkState(cell: Cell, neighbours: Cell[]): Cell {
    const aliveCount = neighbours.filter((cell) => isAlive(cell)).length;
    if (isDead(cell)) {
        const aliveCount = neighbours.filter((cell) => isAlive(cell)).length;
        if (aliveCount === 3) {
            return { ...cell, kind: "Alive" };
        }
        return cell;
    } else {
        if (aliveCount === 2 || aliveCount === 3) {
            return cell;
        }
        return { ...cell, kind: "Dead" };
    }
}


function getRandomInt(min: number, max: number): number {
    min = Math.ceil(min);
    max = Math.floor(max);
    return Math.floor(Math.random() * (max - min)) + min;
}

export function createRandom(coordinate: Coordinate): Cell {
    const num = getRandomInt(1, 100000);
    if (num % 2 === 0) {
        return { kind: "Alive", Coordinate: coordinate };
    }
    return { kind: "Dead", Coordinate: coordinate };
}

File: src/Cell.fs
namespace GameOfLife


[<Struct>]
type Coordinates = | TwoDimensionCoordinate of  x: int * y: int 

type Cell =  
    | Alive
    | Dead

module Cell =
    open System

    let private r = Random()
    
    let random () =
        if r.Next() % 2 = 0 then
            Alive
        else
            Dead
        
    let name cell = 
        match cell with 
        | Alive -> "alive"
        | Dead -> "dead"

    let updateState neighbours cell =
        let aliveCount = neighbours |> Seq.filter(fun n -> match n with | Alive(_) -> true | _ -> false) |> Seq.length
        match cell with
        | Dead when aliveCount = 3 ->
            Alive
        | Alive when not (aliveCount = 2 || aliveCount = 3) -> 
            Dead
        | cell -> cell 

module Coordinates = 
    let isValid (coor: Coordinates) = 
        let (TwoDimensionCoordinate(x, y)) = coor
        x >= 0 && y >= 0

    let isNeighbour coor1 coor2 =
        true
    